src/v/cloud_storage/remote_partition.cc

```diff
@@ -640,7 +635,19 @@ std::unique_ptr<remote_segment_batch_reader> remote_partition::borrow_reader(
           return res;
       },
       [this, &config](materialized_segment_ptr& m_state) {
-          return m_state->borrow_reader(config, _ctxlog, _probe);
+          auto reader = m_state->borrow_reader(config, _ctxlog, _probe);
+          while (m_state->readers.size() > 1) {
```

Comment:
On line 626 `gc_stale_materialized_segments` is called with parameter `force_collection` set to true. This is supposed to remove all existing readers from materialized segment.

This might not work because borrowed readers are not stored inside the `m_state->readers` collections. The method is called by the `partition_record_batch_reader_impl` which stores the reader internally while it's used to fetch data. When the `partition_record_batch_reader_impl` is done with the reader it calls `return_reader` and the reader is getting back into materialized segment. 

Maybe, the good solution would be to just keep track of readers which are outside of the materialized segment right now. And to also impose a limit on max number of readers in existence. In this method if segment is materialized we should probably wait until the reader will be returned back to the materialized segment and then close it and proceed with creation of the new reader.

Reply:
For this looks like we should be applying backpressure. Materilalized segments won't be evicted if the reader that uses it exists. It is possible to have a lot of segments/readers and we can't forcibly stop one which is in use. So we should probably wait  until we will be able to keep only one materialized segment in the partition. Maybe the limit should be not per partition but per shard.